Random Forest

In [14]:
import pandas as pd
from sklearn.ensemble import RandomForestClassifier
from sklearn.model_selection import train_test_split
from sklearn import metrics

from sklearn.feature_selection import RFE 

In [15]:
#col_names = ['age','prescription','astigmatic','tear_rate','lenses']
col_names = ['id', 'diagnosis', 'radius', 'texture', 'perimeter', 'area', 'smoothness', 'compactness', 
             'concavity', 'concave points', 'symmetry', 'fractal dimension', 'se_radius', 'se_texture', 
             'se_perimeter', 'se_area', 'se_smoothness', 'se_compactness', 'se_concavity', 'se_concave points', 
             'se_symmetry', 'se_fractal dimension', 'worst_radius', 'worst_texture', 'worst_perimeter', 
             'worst_area', 'worst_smoothness', 'worst_compactness', 'worst_concavity', 'worst_concave points',
             'worst_symmetry', 'worst_fractal dimension']
dataset = pd.read_csv("../Datasets/wdbc.tab", sep='\t',header=None, names=col_names)

In [16]:
dataset

id  diagnosis      radius     texture   perimeter        area  \
0          id  diagnosis      radius     texture   perimeter        area   
1    discrete   discrete  continuous  continuous  continuous  continuous   
2      ignore      class         NaN         NaN         NaN         NaN   
3      842302          M       17.99       10.38       122.8        1001   
4      842517          M       20.57       17.77       132.9        1326   
..        ...        ...         ...         ...         ...         ...   
567    926424          M       21.56       22.39         142        1479   
568    926682          M       20.13       28.25       131.2        1261   
569    926954          M        16.6       28.08       108.3       858.1   
570    927241          M        20.6       29.33       140.1        1265   
571     92751          B        7.76       24.54       47.92         181   

     smoothness  compactness   concavity  concave points  ...  worst_radius  \
0    smoothness  compactness   concavity  concave points  ...  worst_radius   
1    continuous   continuous  continuous      continuous  ...    continuous   
2           NaN          NaN         NaN             NaN  ...           NaN   
3        0.1184       0.2776      0.3001          0.1471  ...         25.38   
4       0.08474      0.07864      0.0869         0.07017  ...         24.99   
..          ...          ...         ...             ...  ...           ...   
567       0.111       0.1159      0.2439          0.1389  ...         25.45   
568      0.0978       0.1034       0.144         0.09791  ...         23.69   
569     0.08455       0.1023     0.09251         0.05302  ...         18.98   
570      0.1178        0.277      0.3514           0.152  ...         25.74   
571     0.05263      0.04362           0               0  ...         9.456   

     worst_texture  worst_perimeter  worst_area  worst_smoothness  \
0    worst_texture  worst_perimeter  worst_area  worst_smoothness   
1       continuous       continuous  continuous        continuous   
2              NaN              NaN         NaN               NaN   
3            17.33            184.6        2019            0.1622   
4            23.41            158.8        1956            0.1238   
..             ...              ...         ...               ...   
567           26.4            166.1        2027             0.141   
568          38.25              155        1731            0.1166   
569          34.12            126.7        1124            0.1139   
570          39.42            184.6        1821             0.165   
571          30.37            59.16       268.6           0.08996   

     worst_compactness  worst_concavity  worst_concave points  worst_symmetry  \
0    worst_compactness  worst_concavity  worst_concave points  worst_symmetry   
1           continuous       continuous            continuous      continuous   
2                  NaN              NaN                   NaN             NaN   
3               0.6656           0.7119                0.2654          0.4601   
4               0.1866           0.2416                 0.186           0.275   
..                 ...              ...                   ...             ...   
567             0.2113           0.4107                0.2216           0.206   
568             0.1922           0.3215                0.1628          0.2572   
569             0.3094           0.3403                0.1418          0.2218   
570             0.8681           0.9387                 0.265          0.4087   
571            0.06444                0                     0          0.2871   

     worst_fractal dimension  
0    worst_fractal dimension  
1                 continuous  
2                        NaN  
3                     0.1189  
4                    0.08902  
..                       ...  
567                  0.07115  
568                  0.06637  
569                   0.0782  
570                    0.124  
571   

In [17]:
from sklearn.preprocessing import LabelEncoder

In [18]:
X = dataset.iloc[3:, 2:].values
y = dataset.iloc[3:, 1].values

In [19]:
from sklearn.preprocessing import LabelEncoder
labelencoder_y = LabelEncoder()
y = labelencoder_y.fit_transform(y)

In [20]:
y

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 1, 1, 0, 0, 0, 0, 1, 0, 1, 1,
       0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1,
       0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 1, 0, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 0, 1, 1,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0,
       0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1,

In [21]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.3)

In [22]:
clf = RandomForestClassifier(max_depth=5, n_estimators=10, max_features=1)
#clf = RandomForestClassifier()
clf = RFE(clf, 3, step = 1)

In [23]:
clf = clf.fit(X_train, y_train)
y_pred = clf.predict(X_test)

In [24]:
y_pred

array([1, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1,
       1, 0, 0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 1, 0, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 1, 0,
       0, 0, 0, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0])

In [25]:
y_test

array([1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 1, 0, 1,
       1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1,
       1, 0, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0])

In [26]:
print("Accuracy:",metrics.accuracy_score(y_test, y_pred))

Accuracy: 0.9181286549707602


In [27]:
clf.ranking_

array([10, 14, 16,  1, 15, 22, 24,  1, 28, 25,  6, 21,  9, 19, 13,  8, 12,
       27, 26, 17,  3,  7,  4, 11, 20, 18,  1,  2,  5, 23])

COMMENTS:

Using Orange3 for this dataset gave an accuracy of 95% on the test set, whereas in Scikit Learn it gave an accuracy of 91.8% with the same hyper parameter settings which can be taken as approximately same.

The most significant features in the dataset were found to be:

    1. worst_concavity 
    2. worst_concave points
    3. worst_radius
    4. worst_perimeter
    5. worst_symmetry